### DISCUSSION TITLE
v0.4 beta 1 - Please test and report issues

### DISCUSSION THREAD

**Discussion by jschwe** [corrosion-rs/corrosion#389] [Announcements]
# 0.4.0-beta1 (2023-05-10)

Please test and provide feedback for issues you encounter. We are especially interested in receiving feedback for the things in the `experimental` section of the release notes, especially the cbindgen and cxxbridge integration.

Changes compared to v0.3.5:

## Breaking Changes

- The Visual Studio Generators now require at least CMake 3.20.
  This was previously announced in the 0.3.0 release notes and is the same
  requirement as for the other Multi-Config Generators.
- The previously deprecated function `corrosion_set_linker_language()`
  will now raise an error when called and may be removed without further
  notice in future stable releases. Use `corrosion_set_linker()` instead.
- Improved the FindRust target triple detection, which may cause different behavior in some cases.
  The detection does not require an enabled language anymore and will always fall back
  to the default host target triple. A warning is issued if target triple detection failed.

## Potentially Breaking Changes

- Corrosion now sets the `IMPORTED_NO_SONAME` property for shared rust libraries, since by
  default they won't have an `soname` field.
  If you add a rustflag like `-Clink-arg=-Wl,-soname,libmycrate.so` in your project,
  you should set this property to false on the shared rust library.
- Corrosion now uses a mechanism to determine which native libraries need to be linked with
  Rust `staticlib` targets into C/C++ targets. The previous mechanism contained a hardcoded list.
  The new mechanism asks `rustc` which libraries are needed at minimum for a given
  target triple (with `std` support). This should not be a breaking change, but if you
  do encounter a new linking issue when upgrading with `staticlib` targets, please open an
  issue.

## New features

- `corrosion_import_crate()` has two new options `LOCKED` and `FROZEN` which pass the 
  `--locked` and `--frozen` flags to all invocations of cargo.
- `FindRust` now provides cache variables containing information on the default host
  target triple:
  - `Rust_CARGO_HOST_ARCH`
  - `Rust_CARGO_HOST_VENDOR`
  - `Rust_CARGO_HOST_OS`
  - `Rust_CARGO_HOST_ENV`

## Other changes

- When installing Corrosion with CMake >= 3.19, the legacy Generator tool is
  no longer built and installed by default.
- Corrosion now issues a warning when setting the linker or setting linker
  options for a Rust static library.
- Corrosion no longer enables the `C` language when CMake is in crosscompiling mode and
  no languages where previously enabled. This is not considered a breaking change.
- `corrosion_import_crate()` now warns about unexpected arguments.

## Fixes

- Fix building when the `dev` profile is explicitly set by the user.

## Experimental status (may be changed or removed before a stable release)
- Experimental cxxbridge and cbindgen integration.
- Add a helper function to parse the package version from a Cargo.toml file
- Expose rustup toolchains discovered by `FindRust` in the following cache variables
  which contain a list.
  - `Rust_RUSTUP_TOOLCHAINS`: List of toolchains names
  - `Rust_RUSTUP_TOOLCHAINS_VERSION`: List of `rustc` version of the toolchains
  - `Rust_RUSTUP_TOOLCHAINS_RUSTC_PATH`: List of the path to `rustc`
  - `Rust_RUSTUP_TOOLCHAINS_CARGO_PATH`: List of the path to `cargo`. Entries may be `NOTFOUND` if cargo
    is not available for that toolchain.
- Add target properties `INTERFACE_CORROSION_RUSTC` and `INTERFACE_CORROSION_CARGO`, which may
  be set to paths to `rustc` and `cargo` respectively to override the toolchain for a specific
  target.

<hr /><em>This discussion was created from the release <a href='https://github.com/corrosion-rs/corrosion/releases/tag/v0.4-beta1'>v0.4 beta 1 - Please test and report issues</a>.</em>

**Comment by ogoffart** [comment]
At least, our CI passed with v0.4 beta: https://github.com/slint-ui/slint/pull/2731  🎉

**Comment by OlivierLDff** [comment]
Our ci using cxxbridge function works like a charm on Ubuntu (ninja) and windows (msvc)

**Reply by jschwe** [reply]
@OlivierLDff How is your general experience with the cxxbridge function? We don't use it at work (no C++ in our codebase), so the testing on my side is quite limited. Currently the function is documented as experimental and I'm interested in hearing real word usage experiences before commiting to stabilizing the current API of the function.

**Reply by OlivierLDff** [reply]
So we are using it to add some rust to our C++ codebase.

Usage is pretty simple, no special hack or advance usage (which is nice, for once)

We have 2 crates in a workspace called `kapta-rs` and another `kapta-cxx` that do the binding of rust code to C++ using `cxx` library.

```cmake
set(KAPTA_TARGET
    "kapta_rs"
    CACHE STRING "Target Name"
)

if(KAPTA_BUILD_SHARED)
  set(KAPTA_CXX_CRATE_TYPE cdylib)
else()
  set(KAPTA_CXX_CRATE_TYPE staticlib)
endif()

corrosion_import_crate(
  MANIFEST_PATH src/kapta-cxx/Cargo.toml
  IMPORTED_CRATES KAPTA_CXX_CRATE
  CRATE_TYPES ${KAPTA_CXX_CRATE_TYPE}
  CRATES kapta-cxx
)

corrosion_add_cxxbridge(
  ${KAPTA_TARGET}
  CRATE ${KAPTA_CXX_CRATE}
  FILES lib.rs
)
```

I import corrosion using [CPM.cmake](https://github.com/cpm-cmake/CPM.cmake)

```cmake
# Heuristic to detect if corrosion is already imported
if(COMMAND corrosion_import_crate)
  return()
endif()

include(${CMAKE_CURRENT_LIST_DIR}/CPM.cmake)

CPMAddPackage(
  NAME corrosion
  GIT_REPOSITORY "https://github.com/Naostage/corrosion"
  GIT_TAG "4c36abff079364d6ebdf51526fc3457c216a0545"
)
```

**Comment by temeddix** [comment]
We're getting errors. We have CMake 3.25 installed, but the error tells us that we need to have CMake 3.20 or higher.
https://github.com/cunarist/app-template/discussions/25#discussion-5214646

**Reply by jschwe** [reply]
I already replied in the linked issue, but for completeness sake: this is a user issue, where a vendored CMake version provided by the IDE is used, which is less recent than the installed CMake version. Maybe we could print the actual CMake version in the error message, to give the user a better hint at which CMake version specifically is being used.

**Reply by temeddix** [reply]
Yeah it is, I think it would be great to add that info in the error log

**Reply by OlivierLDff** [reply]
Could be interesting to also log full path of used cmake executable? So its even more obvious to the user when non system cmake is used

**Reply by jschwe** [reply]
Thanks -  added in https://github.com/corrosion-rs/corrosion/pull/393
